sklearn/ensemble/_hist_gradient_boosting/_predictor_bitset.pxd

```diff
@@ -0,0 +1,27 @@
+# cython: cdivision=True
+# cython: boundscheck=False
+# cython: wraparound=False
+# cython: language_level=3
+# cython: nonecheck=False
+# distutils: language=c++
+
+from libcpp.set cimport set
+from libcpp.map cimport map
+from libcpp.vector cimport vector
+from .common cimport BITSET_INNER_DTYPE_C
+from .common cimport X_BINNED_DTYPE_C
+from .common cimport X_DTYPE_C
+
+cdef class PredictorBitSet:
+    cdef map[int, vector[BITSET_INNER_DTYPE_C]] node_to_raw_bitset
```

Comment:
I'm a bit concerned about the use of maps. It seems inefficient and unintuitive to hash the nodes indices.

Isn't there a nicer way to have the bitset as an attribute of PREDICTOR_RECORD_DTYPE?

Or should we consider a different approach and redefine the `nodes` array as an array of pure C++ structs?

It seems that what we want to do is easy in pure Python or in pure C++, but very hard in Cython.

Reply:
Wrapping in the predictor data in numpy arrays has the nice side benefit to provide easy pickle support + mmap-ability, for instance with `joblib.dump` / `joblib.load(filename, mmap_mode="r")` which makes it possible to use a model in memory only once for several python processes running on the same host (e.g. the gunicorn workers of a flask application serving scikit-learn models for instance).

Reply:
Could you take a look at https://github.com/scikit-learn/scikit-learn/pull/18140 @ogrisel? We're gearing towards not using numpy structured arrays for this